### ISSUE TITLE
Memory Error while reading for large files

### ISSUE THREAD

**Issue by KarthikRevanuru** [krishnanlab/PecanPy#20]
I've a file with nearly 70M edges and it fails to load into memory on a machine with 32 GB ram

**Comment by KarthikRevanuru** [comment]
@RemyLau

**Comment by KarthikRevanuru** [comment]
I've a graph with 4.5M nodes and 80M edges. What is your rough estimate of running time on large graphs like this.

**Comment by RemyLau** [comment]
Hi @KarthikRevanuru , from my rough estimation, it should take no more than 20GB of memory to fully load and convert your graph into the CSR format, which is used as the final graph data structure. I have run some testing with a couple large biological networks (see [bench repo](https://github.com/krishnanlab/PecanPy_benchmarks)). For example, SSN has roughly 72M edges with 800k nodes, and it uses ~10GB of memory throughout the execution of the program (see line 65 or line 71 in [this](https://github.com/krishnanlab/PecanPy_benchmarks/blob/main/result/stat_summary.tsv#L65) benchmarking result table).

May I ask what mode of execution you are using (i.e. did you explicitly set `--mode` to `PreComp` or `DenseOTF`)?  In this case, since the network has a large number of nodes with very sparse connections, it is best to use `SparseOTF` (which is the default mode).

**Comment by KarthikRevanuru** [comment]
Thanks for quick revert @RemyLau 
I'm using SparseOTF and I've used 32 GB ram but it throws error. On disk my input data size is 3.4 GB, it's stored as edge list.
I've taken a bigger instance and it seems to work. But it has printed only this since 6 hours.
`Took 00:06:06.35 to load graph
Took 00:00:00.00 to pre-compute transition probabilities`

Do you have any estimates on running time ? Also I've enabled verbose mode and it didn't print anything regarding walks

**Comment by RemyLau** [comment]
Hmm.. What error are you seeing? It would be helpful if you can share the error log here. 

From the log message you shared here, it only took 6 minutes to load the graph, which is a quite reasonable time for networks of this size. And since you're using `SparseOTF`, there's no preprocessing step, so 0 sec of preprocessing is expected. But I'm not sure why you only see this after 6 hours. If you look at the command-line interface, there's no other steps before loading graph.

https://github.com/krishnanlab/PecanPy/blob/6a0a73312b99ec476a4d397159becf5e210ab48f/src/pecanpy/cli.py#L202-L233

**Comment by KarthikRevanuru** [comment]
Error message is killed, after I took 64gb ram its fixed.
@RemyLau this is not printed after 6 hrs. Its printed within 6 min of running, but nothing after that since 6 hrs..

**Comment by KarthikRevanuru** [comment]
what's the expected time to generate walks and train embeddings

**Comment by RemyLau** [comment]
@KarthikRevanuru It honestly depends on a lot of factors, e.g. number of processors, CPU clock, memory clock, etc. But in your case, I'll say 6 hours would be the time where the random walk generation process is finished. So given these couple of clues, I think the issue might be caused by the large number of random walks generated. Previously in my case, although SSN network has roughly the same number of edges as yours, it has an order of magnitude less number of nodes (800k compared to 4.5M). The number of nodes does not affect the size of the sparse graph structure much, but it does affect the _size of the corpurs generated_ (i.e. the random walks).

In particular, your `killed` error message may very likely be caused by exceeding memory usage from this line of code in the `simulate_walks` function, which convert the node index sequence into list of strings (of node ids).

https://github.com/krishnanlab/PecanPy/blob/6a0a73312b99ec476a4d397159becf5e210ab48f/src/pecanpy/node2vec.py#L149

This was done originally due to convenience when calling `gensim.Word2Vec`. I'll try to find an alternative solution where we don't need to convert to list of strings first, but instead, use the index sequence directly to save memory usage for specific cases like these (networks with large number of nodes). So keep posted.

In the meantime, if it is possible, try to further increase your memory allocation to, say 128GB, and see if that could resolve the issue.

**Comment by KarthikRevanuru** [comment]
Ok thanks !

**Comment by NavodPeiris** [comment]
@KarthikRevanuru have you tried sharding the edg file into smaller ones and training iteratively?


    edg_files = get_edg_files_in_directory()

    # Initialize the FastText model
    model = FastText(
        vector_size=128, window=10, min_count=1, workers=4, sg=1, hs=1, seed=42, bucket=100000
    )

    first_file = True
    for edg_file in edg_files:
        print(f"Processing {edg_file}...")
        
        # Load graph and simulate walks
        g.read_edg(edg_file, weighted=True, directed=False, delimiter="‖")
        walks = g.simulate_walks(num_walks=200, walk_length=20)
        
        if first_file:
            # Build vocabulary from the first batch of walks
            model.build_vocab(walks)
            first_file = False
        else:
            # Update vocabulary incrementally
            model.build_vocab(walks, update=True)
        
        # Train model incrementally
        model.train(walks, total_examples=len(walks), epochs=1)
    
        print("Saving the model...")
        # Convert word vectors to float16 to reduce memory usage
        model.wv.vectors = model.wv.vectors.astype(np.float16)

        # Save the model with reduced precision
        model.save("graph_embeddings.model")
